File: src/terminal.rs
use std::io::{self, Write};
use std::mem::MaybeUninit;

pub struct Terminal {
    original_termios: libc::termios,
}

impl Terminal {
    pub fn new() -> Self {
        let mut original = unsafe { MaybeUninit::<libc::termios>::zeroed().assume_init() };
        unsafe { libc::tcgetattr(libc::STDIN_FILENO, &mut original) };

        let mut term = Self {
            original_termios: original,
        };
        term.enable_raw_mode();
        term.hide_cursor();
        term
    }

    fn enable_raw_mode(&mut self) {
        let mut raw = self.original_termios;
        raw.c_lflag &= !(libc::ECHO | libc::ICANON);
        raw.c_cc[libc::VMIN] = 0;
        raw.c_cc[libc::VTIME] = 0;
        unsafe { libc::tcsetattr(libc::STDIN_FILENO, libc::TCSAFLUSH, &raw) };
    }

    pub fn hide_cursor(&self) {
        print!("\x1B[?25l");
        let _ = io::stdout().flush();
    }

    pub fn show_cursor(&self) {
        print!("\x1B[?25h");
        let _ = io::stdout().flush();
    }

    pub fn move_to(&self, row: usize, col: usize) {
        print!("\x1B[{row};{col}H");
        let _ = io::stdout().flush();
    }

    pub fn clear_screen(&self) {
        print!("\x1B[2J");
        let _ = io::stdout().flush();
    }

    pub fn key_pressed(&self) -> bool {
        let mut c: u8 = 0;
        unsafe {
            libc::read(
                libc::STDIN_FILENO,
                &mut c as *mut u8 as *mut libc::c_void,
                1,
            ) == 1
        }
    }

    pub fn size() -> (usize, usize) {
        let mut w = unsafe { MaybeUninit::<libc::winsize>::zeroed().assume_init() };
        unsafe { libc::ioctl(libc::STDOUT_FILENO, libc::TIOCGWINSZ, &mut w) };
        (w.ws_row as usize, w.ws_col as usize)
    }
}

impl Default for Terminal {
    fn default() -> Self {
        Self::new()
    }
}

impl Drop for Terminal {
    fn drop(&mut self) {
        self.show_cursor();
        unsafe {
            libc::tcsetattr(libc::STDIN_FILENO, libc::TCSAFLUSH, &self.original_termios);
        }
    }
}

File: src/terminal.c
#include <termios.h>
#include <unistd.h>
#include <sys/ioctl.h>

static struct termios orig_termios;
static int raw_mode_enabled = 0;

void term_enable_raw(void) {
    if (raw_mode_enabled) return;
    tcgetattr(STDIN_FILENO, &orig_termios);
    struct termios raw = orig_termios;
    raw.c_lflag &= ~(ECHO | ICANON);
    raw.c_cc[VMIN] = 0;
    raw.c_cc[VTIME] = 0;
    tcsetattr(STDIN_FILENO, TCSAFLUSH, &raw);
    raw_mode_enabled = 1;
}

void term_disable_raw(void) {
    if (!raw_mode_enabled) return;
    tcsetattr(STDIN_FILENO, TCSAFLUSH, &orig_termios);
    raw_mode_enabled = 0;
}

int term_key_pressed(void) {
    char c;
    return read(STDIN_FILENO, &c, 1) == 1;
}

void term_get_size(int *rows, int *cols) {
    struct winsize w;
    if (ioctl(STDOUT_FILENO, TIOCGWINSZ, &w) == 0) {
        *rows = w.ws_row;
        *cols = w.ws_col;
    } else {
        *rows = 24;
        *cols = 80;
    }
}

void term_hide_cursor(void) {
    write(STDOUT_FILENO, "\033[?25l", 6);
}

void term_show_cursor(void) {
    write(STDOUT_FILENO, "\033[?25h", 6);
}
